### PR TITLE
base_url support

### PR DESCRIPTION
**Pull request by haraldschilly**
issue #69

This is my probably quite naive approach to introduce a `base_url`. I wasn't really able to test it, but I think it should do it's job. The basic idea is to ignore a specific prefix for incoming request paths. In particular, running it via

     ./bazel-bin/tensorboard/tensorboard --logdir=~/tmp --base_url=/foo/bar

let's me work with the tensorboard at

    http://localhost:6006/foo/bar/#graphs

I think the only really relevant code is in `backend/application.py` → `TensorBoardWSGIApp::__call__`. I think it's safeguarded in a way such that any other (non-base_url-prefixed) paths still work.

### CODE REVIEW

**Comment by jart** [comment]
Thank you for your contribution.

I'm not sure if this is something we would want to maintain in the long run. It might be more appropriate to configure this at the systems level.

For example, would NGINX be able to solve your use case? Here is how such a thing could be configured using the [rewrite module](http://nginx.org/en/docs/http/ngx_http_rewrite_module.html):

```nginx
server {
  listen 80 default_server;
  listen [::]:80 default_server ipv6only=on;

  location = /tensorboard {
    return 302 /tensorboard/;
  }

  location /tensorboard {
    rewrite ^/tensorboard(/.*)$ $1 break;
    proxy_http_version 1.1;
    proxy_pass http://127.0.0.1:6006;
  }
}
```

I can now visit http://localhost/tensorboard/ in the browser and it will d

**Comment by haraldschilly** [comment]
Hi, I tried to merge/update this, but it still fails. I'll try again another day, and I understand I should also add a test case.

However, in the case I deal with there is no provision for such a http request rewrite. Unrelated but similar to this, there is also a browser-based application called "jupyter notebook". It does [also support this base_url setting](http://jupyter-notebook.readthedocs.io/en/latest/config.html), so I assumed it's within scope here, too.

**Comment by JGSweets** [comment]
@jart Unfortunately that does not fix the problem. For instance, if the path is routed through an ELB i.e.
http://youraddress.com/abc/tensorboard

While you can reach the website, the logs cannot be attained.
http://youraddress.com/data/logdir
is requested when it actually lies in
http://youraddress.com/abc/data/logdir

**Comment by JGSweets** [comment]
@haraldschilly any update?

also noticed that many of the fails were due to the restructuring of  TensorBoardWSGIApp in the tests.

TensorBoardWSGIApp() takes exactly 5 arguments (4 given)

You restructured  TensorBoardWSGIApp, but did not update the tests to handle the altered function.

**Comment by jart** [comment]
> @jart Unfortunately that does not fix the problem. For instance, if the path is routed through an ELB i.e. http://youraddress.com/abc/tensorboard While you can reach the website, the logs cannot be attained. http://youraddress.com/data/logdir is requested when it actually lies in http://youraddress.com/abc/data/logdir

I'm afraid I don't understand. If NGINX is used to put TensorBoard under the base path /abc/tensorboard/ then logdir would be available under /abc/tensorboard/data/logdir.

**Comment by JGSweets** [comment]
@jart 
It's elb -> nginx -> TB

Unfortunately I can't use subdomains so the elb (path based routing) is the domain. The elb routes to nginx through the path /abc. Hence, the client side of TB requests youraddress.com and the ELB won't know where to route it.

File: tensorboard/backend/application.py [hunk 1]
```diff
@@ -64,7 +64,8 @@ def standard_tensorboard_wsgi(
     reload_interval,
     plugins,
     db_uri="",
-    assets_zip_provider=None):
+    assets_zip_provider=None,
+    base_url=""):
```
Comment (jart) [inline tensorboard/backend/application.py]: Please make this `base_prefix='/'`

File: tensorboard/backend/application.py [hunk 2]
```diff
@@ -92,17 +94,18 @@ def standard_tensorboard_wsgi(
       assets_zip_provider=(assets_zip_provider or
                            get_default_assets_zip_provider()))
   plugins = [constructor(context) for constructor in plugins]
-  return TensorBoardWSGIApp(logdir, plugins, multiplexer, reload_interval)
+  return TensorBoardWSGIApp(logdir, plugins, multiplexer, reload_interval, base_url)
 
 
-def TensorBoardWSGIApp(logdir, plugins, multiplexer, reload_interval):
+def TensorBoardWSGIApp(logdir, plugins, multiplexer, reload_interval, base_url):
```
Comment (jart) [inline tensorboard/backend/application.py]: `base_prefix='/'`

File: tensorboard/main.py [hunk 3]
```diff
@@ -108,6 +108,9 @@
     'event_file', '',
     'The particular event file to query for. Only used if --inspect is present '
     'and --logdir is not specified.')
+tf.flags.DEFINE_string(
+    'base_url', '',
```
Comment (jart) [inline tensorboard/main.py]: `'base_prefix', '/'`

File: tensorboard/backend/application.py [hunk 4]
```diff
@@ -386,8 +391,10 @@ def _get_connect_params(query):
   return {k: v[0] for k, v in params.items()}
 
 
-def _clean_path(path):
-  """Removes trailing slash if present, unless it's the root path."""
+def _clean_path(path, _base_url = ""):
+  """Removes _base_url part and trailing slash if present, unless it's the root path."""
+  if len(_base_url) > 0 and clean_path.startswith(_base_url):
```
Comment (jart) [inline tensorboard/backend/application.py]: Shouldn't we 404 if path doesn't start with the prefix instead?

**Comment by JGSweets** [comment]
Since I do not have access to commiting to this PR, I suggest closing this PR as I will be submitting a PR which resolves this issue.

**Comment by chihuahua** [comment]
Closing because #436 got merged.

**Pull request closed without merging**